File: 0445-add-two-numbers-ii/0445-add-two-numbers-ii.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val = 0, _next = nil)
#         @val = val
#         @next = _next
#     end
# end
# @param {ListNode} l1
# @param {ListNode} l2
# @return {ListNode}
def add_two_numbers(l1, l2)
    r = ''
    s = ''
    while l1
        r += l1.val.to_s
        l1 = l1.next
    end
    while l2
        s += l2.val.to_s
        l2 = l2.next
    end
    r = (r.to_i+s.to_i).to_s.chars.map(&:to_i).map {|x| ListNode.new(x)}.to_a
    for i in (1..r.length)
        r[i-1].next = r[i]
    end
    r[0]
end

File: 0445-add-two-numbers-ii/0445-add-two-numbers-ii.php
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn add_two_numbers(l1: Option<Box<ListNode>>, l2: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut v1 = Solution::list2vec(&l1);
        let mut v2 = Solution::list2vec(&l2);
        v1.reverse();
        v2.reverse();
        let mut next = None;
        let mut carry = 0;
        for i in 0..=std::cmp::max(v1.len(), v2.len()) {
            let mut val =
                carry + if i < v1.len() { v1[i] } else { 0 } + if i < v2.len() { v2[i] } else { 0 };
            carry = if val > 9 {
                val -= 10;
                1
            } else {
                0
            };
            if i < std::cmp::max(v1.len(), v2.len()) || val != 0 {
                next = Some(Box::new(ListNode { val, next }));
            }
        }
        next
    }
    fn list2vec(list: &Option<Box<ListNode>>) -> Vec<i32> {
        if list.is_none() {
            return vec![0];
        }
        let mut ret = Vec::new();
        let mut node = list;
        while let Some(n) = node {
            ret.push(n.val);
            node = &n.next;
        }
        ret
    }
}